Web Inspector: Performance issue with search tab and single character

rcaliman-apple opened the pull request:
#### a2e3ab02b806f1bb8532d598b96ff123daaf03f0
<pre>
Web Inspector: Performance issue with search tab and single character
<a href="https://rdar.apple.com/49234522">rdar://49234522</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=262158">https://bugs.webkit.org/show_bug.cgi?id=262158</a>

Reviewed by Devin Rousso.

Web Inspector search works by first requesting a list of all resources
that have matches for a query. Then, iterating through each resource,
it gets lines of text where matches are found. Finally the same search
query regex is applied to each line to get individual search result matches.

Before this patch, a `WI.SearchResultTreeElement` was created and attached
for each match, regardless of how many. This was expensive for a few reasons:

- for each result, a few substring operations are run on the line of text
to provide context around the individual match. On minified files,
this means working on almost the entire contents of the file.

- attaching the tree element to the tree outline creates a DOM structure for each element,
without necessarily attaching it to the DOM (tree outline virtualization handles that),
adds event handlers, creates relationships with sibling tree elements, etc.
All this work is done regardless if the tree element will ever be seen.

For queries with a very large number of results (tens or hundreds of thousands),
there&apos;s a lot unnecessary work done before even showing the first few results.
This work occurs on the UI thread and blocks interaction within Web Inspector.

From a UX perspective, a user overwhelmed with a large volume of search results
is more likely to want to refine their query than see all of the results.

This patch reduces the amount of work done:
- for each resource, show only the first 100 results by default.

- provide UI controls per resource to load more results or all results.
This allows a user to intentionally request to show a known large number of results
with the implied acknowledgement that the action might cause a slowdown.
This follows a pattern used elsewhere in Web Inspector: showing contents of Arrays,
Sets, Maps, or DataGrids with many entries.

- other piecemeal optimizations:
-- event delegation for handling double-click on individual search result tree elements

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.css: Added.
(.tree-outline .item.search-results-placeholder &gt; button):
(&amp;:first-of-type):
* Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js: Added.
(WI.SearchResultsPlaceholderTreeElement):
(WI.SearchResultsPlaceholderTreeElement.prototype.onattach):
(WI.SearchResultsPlaceholderTreeElement.prototype._handleShowMoreButtonClicked):
(WI.SearchResultsPlaceholderTreeElement.prototype._handleShowAllButtonClicked):
* Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
(WI.SearchSidebarPanel):
(WI.SearchSidebarPanel.prototype.performSearch):
(WI.SearchSidebarPanel.prototype.attached):
(WI.SearchSidebarPanel.prototype.detached):
(WI.SearchSidebarPanel.prototype.matchTreeElementAgainstCustomFilters):
(WI.SearchSidebarPanel.prototype.searchResultsPlaceholderLoadMoreResults):
(WI.SearchSidebarPanel.prototype._renderResultsForSourceCodeTreeElement):
(WI.SearchSidebarPanel.prototype._createSearchResultsPlaceholderTreeElementIfNeeded):
(WI.SearchSidebarPanel.prototype.performSearch.createTreeElementForMatchObject): Deleted.
(WI.SearchSidebarPanel.prototype._treeElementDoubleClick): Deleted.
* Source/WebInspectorUI/UserInterface/Views/TreeElement.js:
(WI.TreeElement.prototype._setListItemNodeContent):
(WI.TreeElement.prototype.get titleHTML): Deleted.
(WI.TreeElement.prototype.set titleHTML): Deleted.
Drive-by clean-up of unused `titleHTML`

Canonical link: <a href="https://commits.webkit.org/306429@main">https://commits.webkit.org/306429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8c1513663a29766e9ed01e00fa1284c76cde64f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/141342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/13726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/3052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/149920 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/94441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/14441 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/13881 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/108596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/94441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/144293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/11145 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/126496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/89501 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/10719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/8337 "Passed tests") | [⏳ 🛠 wpe-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/119985 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/2471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/152310 "Built successfully") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/2148 "Build is in progress. Recent messages:") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/13417 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/169/builds/2918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/116701 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/13433 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/11723 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/117034 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29788 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/13083 "Passed tests") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/123146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/68615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/13459 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/2498 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/13196 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/77172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/13396 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/13242 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -2549,10 +2549,12 @@ WI.setZoomFactor = function(factor)
 
 WI.resolvedLayoutDirection = function()
```
Without caching the result, this makes a call over to C++ for every rendered search result.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/TreeElement.js:
```diff
@@ -108,18 +108,6 @@ WI.TreeElement = class TreeElement extends WI.Object
         this.didChange();
     }
 
-    get titleHTML()
-    {
-        return this._titleHTML;
-    }
-
-    set titleHTML(x)
-    {
-        this._titleHTML = x;
-        this._setListItemNodeContent();
-        this.didChange();
-    }
-
```
These aren't use anywhere.

dcrousso replied:
then we should also remove the references to it in `_setListItemNodeContent` :)

rcaliman-apple replied:
Indeed. Removed 👍

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -2549,10 +2549,12 @@ WI.setZoomFactor = function(factor)
-    return layoutDirection;
+    if (!WI._layoutDirection) {
+        let layoutDirection = WI.settings.debugLayoutDirection.value;
+        WI._layoutDirection = layoutDirection === WI.LayoutDirection.System ? InspectorFrontendHost.userInterfaceLayoutDirection() : layoutDirection;
```
what if the layout direction changes (either by macOS itself or the debug setting)?

rcaliman-apple replied:
When the debug setting is changed, Web Inspector is reopens automatically:
https://github.com/WebKit/WebKit/blob/9e3909849159cacd82f6d7cf78b76fde38038fca/Source/WebInspectorUI/UserInterface/Base/Main.js#L2582

There's no event emitted from the backend when RTL/LTR changes at the system level. Even today (if a system restart isn't required; didn't try it), Web Inspector doesn't handle that. The UI would be inconsistent until closing and reopening it, which I think is a viable workaround for an action that is seldom used.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultTreeElement.js:
```diff
@@ -50,25 +52,29 @@ WI.SearchResultTreeElement = class SearchResultTreeElement extends WI.GeneralTre
+        }
 
-        modifiedTitle = modifiedTitle.truncateEnd(searchTermIndex + searchTermLength + charactersToShowAfterSearchMatch);
+        this._modifiedTitle = prefix + title.substring(trimStartIndex, trimEndIndex) + suffix;
```
this is a `static` method, so referencing `this` is not correct

could we instead just have the caller get the `.textContent` of the result since it seems like is equivalent (though in different DOM structure) to the `modifiedTitle` string?

rcaliman-apple replied:
You're right. We can actually reference `this.mainTitle` which is where the modified string ends up on the superclass, `WI.GeneralTreeElement`.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultTreeElement.js:
```diff
@@ -77,7 +83,7 @@ WI.SearchResultTreeElement = class SearchResultTreeElement extends WI.GeneralTre
     get filterableData()
     {
-        return {text: [this.representedObject.title]};
+        return {text: this._modifiedTitle};
```
i think this needs to remain an array

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+WI.SearchResultsPlaceholderTreeElement = class SearchResultsPlaceholderTreeElement extends WI.TreeElement
```
missing copyright

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+{
+    constructor(delegate, options = {})
+    {
+        super("");
```
style: please use a `const title = "";` so that it's clear from the callsite what the argument is for instead of having to go find out by reading `WI.TreeElement`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+
+        this._delegate = delegate;
+    }
+
```
missing `// Public`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+        this._delegate = delegate;
+    }
+
+    set incrementSize(value) { 
```
style: `{` on separate line

ditto below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+        if (!this._remainingSize || this._remainingSize <= this._incrementSize)
+            this._showAllButtonElement?.remove();
+    }
+
```
missing `// Protected`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+            this._showMoreButtonElement = document.createElement("button");
+            this._showMoreButtonElement.textContent = WI.UIString("Show %d More").format(this._incrementSize);
+            this._showMoreButtonElement.addEventListener("click", this._handleShowMoreButtonClicked.bind(this));
+            this._listItemNode.appendChild(this._showMoreButtonElement);
```
please combine this with the above line for simplicity
```js
this._showMoreButtonElement = this._listItemNode.appendChild(document.createElement("button"));
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+            this._showAllButtonElement = document.createElement("button");
+            this._showAllButtonElement.textContent = WI.UIString("Show All (%d More)").format(this._remainingSize);
+            this._showAllButtonElement.addEventListener("click", this._handleShowAllButtonClicked.bind(this));
+            this._listItemNode.appendChild(this._showAllButtonElement);
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+
+    _handleShowMoreButtonClicked(event)
+    {
+        console.assert(this.parent, "Search results placeholder element must be attached");
```
it's not possible for this event listener to be attached unless this tree element has been attached to a parent so i dont think this assertion really adds any value :/

also, whenever we do have assertions i strongly recommend having `this` be one of the arguments so that if it does fail we get more than just a generic error message in the console (i.e. _which_ tree element failed instead of that _some_ arbitrary tree element failed)

rcaliman-apple replied:
That makes sense. I'll remove these. Thank you for the additional context about `this` in assertions!

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+
+    _handleShowAllButtonClicked(event)
+    {
+        console.assert(this.parent, "Search results placeholder element must be attached");
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+
+        this._delegate.searchResultsPlaceholderLoadMoreResults?.(this.parent, this._remainingSize);
+    }
+}
```
missing newline at end of file

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -54,10 +54,17 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+
+        this._omitMovingFocusToFirstResult = false;
+        this._resultsIncrementSize = 100;
+        this._pendingSearchMatchObjectsForResourceTreeElement = new WeakMap();
```
Style: no need for `()` when using `new` if no arguments are provided

ditto below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -54,10 +54,17 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        this.contentTreeOutline.addEventListener(WI.TreeOutline.Event.ElementAdded, this._handleTreeElementAdded, this);
+
+        this._omitMovingFocusToFirstResult = false;
+        this._resultsIncrementSize = 100;
```
nothing else ever changes this so id make it a `const` somewhere instead of a member variable

rcaliman-apple replied:
It's used both in `_didRenderResultsForResourceTreeElement()` and in various callbacks within `performSearch()`. I'd keep it declared in one place for clarity, even if it's really a `const`.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -295,38 +286,44 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+
+                    let searchMatchObjects = this._pendingSearchMatchObjectsForResourceTreeElement.get(resourceTreeElement);
+                    if (!searchMatchObjects)
+                        searchMatchObjects = [];
```
you can simplify this by adding `|| []` at the end of the previous line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -130,6 +138,7 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         // Before performing a new search, clear the old search.
         this.contentTreeOutline.removeChildren();
         this.contentBrowser.contentViewContainer.closeAllContentViews();
+        this._pendingSearchMatchObjectsForResourceTreeElement = new WeakMap();
```
should this also reset `this._searchResultsPlaceholderTreeElementForResourceTreeElement`?

rcaliman-apple replied:
Good catch! 👍

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -449,14 +454,71 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
-        } else if (treeElement.representedObject instanceof WI.SourceCodeSearchMatchObject) {
+        }
+
+        if (treeElement.representedObject instanceof WI.SourceCodeSearchMatchObject) {
```
why was this changed?  aren't `WI.DOMSearchMatchObject` and `WI.SourceCodeSearchMatchObject` mutually exclusive (i.e. if `treeElement.representedObject` is an `instanceof` one then it cant be an `instanceof` the other)?

rcaliman-apple replied:
The change was purely esthetic, to remove the `else if`. Yes, the two are mutually exclusive and they remain so even in this form. I'll revert it nonetheless.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -54,10 +54,17 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         this._inputElement.value = this._searchQuerySetting.value;
 
         WI.Frame.addEventListener(WI.Frame.Event.MainResourceDidChange, this._mainResourceDidChange, this);
+        WI.SearchResultTreeElement.addEventListener(WI.TreeElement.Event.DoubleClick, this._handleTreeElementDoubleClick, this);
```
i recognize that this is a very clever solution to avoid having to add a listener for `WI.TreeElement.Event.DoubleClick` on every `WI.SearchResultTreeElement`, but that only works if `WI.SearchResultTreeElement` is not used anywhere else

while that is the reality today, im concerned that this would result in subtle bugs in the future, so id suggest moving this to an `attached()` and removing it in a `detached()` just in case

rcaliman-apple replied:
That's a good idea! 👍

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -2549,10 +2549,12 @@ WI.setZoomFactor = function(factor)
-    if (layoutDirection === WI.LayoutDirection.System)
-        layoutDirection = InspectorFrontendHost.userInterfaceLayoutDirection();
-    return layoutDirection;
+    if (!WI._layoutDirection) {
```
i feel like this should be split into a separate PR

rcaliman-apple replied:
I split this off into https://github.com/WebKit/WebKit/pull/56723

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultTreeElement.js:
```diff
@@ -77,7 +80,7 @@ WI.SearchResultTreeElement = class SearchResultTreeElement extends WI.GeneralTre
     get filterableData()
     {
-        return {text: [this.representedObject.title]};
+        return {text: [this.mainTitle]};
```
i feel like this should be split into a separate PR

rcaliman-apple replied:
I split this off into https://github.com/WebKit/WebKit/pull/56723

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -449,14 +454,71 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         }
     }
 
+     _handleTreeElementAdded(event)
```
can you explain a bit more why this change was made?

is this just to make it so that when the "Show %d More" is clicked that the first tree element added that way becomes selected?  if so, i feel like it'd be easier to just do that in `searchResultsPlaceholderLoadMoreResults` instead of listening for _every_ tree element being added

rcaliman-apple replied:
This is unrelated to the clicks on the "Show More" button. 

It's code moved from the [now removed( `createTreeElementForMatchObject()`](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js#L182-L185) that was called to add every search result tree element. The goal is, yes, to select the first search tree element to reveal its represented object's content.

Logically, it's the same. Syntactically, I felt it's clearer to put this into an event handler.

I can inline it into `_didRenderResultsForResourceTreeElement()` which is called a lot less frequently. Or dynamically add and remove the event handler 🤔

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -449,14 +454,71 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+
+        while (limit) {
+            resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));
+            limit--;
```
NIT: you can move this earlier to be `while (limit--)` to avoid the extra line and `{` `}`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeOutline.js:
```diff
@@ -297,6 +297,44 @@ WI.TreeOutline = class TreeOutline extends WI.Object
             this.expand();
     }
 
+    moveChildToIndex(child, index)
```
NIT: id just call this `moveChild` to better match `insertChild`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeElement.js:
```diff
@@ -255,11 +240,11 @@ WI.TreeElement = class TreeElement extends WI.Object
     }
 
-    _attach()
+    attach()
```
while i agree with your argument that these aren't really private, they're private to `WI.TreeElement` and `WI.TreeOutline` (i.e. not meant to be called by anything else) so id prefer if we kept them as is to avoid some future developer thinking that these can be called by anyone

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeOutline.css:
```diff
@@ -91,6 +91,21 @@ body[dir=rtl] .tree-outline .item :is(.disclosure-button, .icon) {
     margin-inline-start: var(--tree-outline-icon-margin-start);
 }
 
+.tree-outline .item.placeholder-item > button {
```
these should really be moved to a separate `Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.css`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,65 @@
+            this._listItemNode.appendChild(this._showAllButtonElement);
+        }
+
+        this._listItemNode.classList.add("item", "placeholder-item");
```
i think this should really be `"search-placeholder-item"` in case something else comes along that also needs to be a `"placeholder-item"`

either that or rename this to be more generic like `WI.ShowMorePlaceholderTreeElement`

dcrousso reviewed (commented):
very cool!

mostly style/nit comments with a few substantive ones id like to see addressed before approving

overall great work! :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,91 @@
+
+    _handleShowMoreButtonClicked(event)
+    {
+        this._delegate.searchResultsPlaceholderLoadMoreResults?.(this.parent, this._incrementSize);
```
i realize that as it's currently used this will always be a `WI.ResourceTreeElement`, but that is not codified anywhere in this class, so id have it just return `this` instead (which btw is also the general pattern when using delegates)

we can then have `searchResultsPlaceholderLoadMoreResults` fetch and use the `parent`
```js
searchResultsPlaceholderLoadMoreResults(searchResultsPlaceholderTreeElement, limit)
{
    let resourceTreeElement = searchResultsPlaceholderTreeElement.parent;
    console.assert(resourceTreeElement instanceof WI.ResourceTreeElement, resourceTreeElement);

    this._renderResultsForResourceTreeElement(resourceTreeElement, limit);
}
```

rcaliman-apple replied:
Good call. I'll add a check for `WI.ScriptTreeElement` to the assertion

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,91 @@
+            this._showAllButtonElement.addEventListener("click", this._handleShowAllButtonClicked.bind(this));
+        }
+
+        this._listItemNode.classList.add("item", "search-placeholder-item");
```
NIT: so since we already have `"item"` i dont think we need it at the end of `"search-placeholder"` (i.e. `.item.search-placeholder-item` is more redundant than `.item.search-placeholder`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -58,6 +58,11 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         this.contentTreeOutline.addEventListener(WI.TreeOutline.Event.SelectionDidChange, this._treeSelectionDidChange, this);
+
+        this._omitMovingFocusToFirstResult = false;
+        this._resultsIncrementSize = 100;
```
i was thinking more like `WI.SearchSidebarPanel._resultsIncrementSize` as that way it's not a property on every instance of `WI.SearchSidebarPanel` (although i recognize that there's only ever one instance right now)

rcaliman-apple replied:
Yes, that makes sense.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -249,23 +244,16 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
-
-            for (let script of scriptsToSearch)
-                countPromise(script.target.DebuggerAgent.searchInContent(script.id, searchQuery, isCaseSensitive, isRegex), scriptCallback.bind(this, script));
+            this._renderResultsForResourceTreeElement(scriptTreeElement, this._resultsIncrementSize);
```
i think we should rename this `_renderResultsForSourceCodeTreeElement` since we're giving it a `WI.ScriptTreeElement` here

rcaliman-apple replied:
Even if we create an intermediary `_renderResultsForSourceCodeTreeElement` for clarity, that will still call `_renderResultsForResourceTreeElement`. There's no point in duplicating the logic. We only care that the thing is a `WI.TreeElement` to append children to it.

(We also use a `WI.ResourceTreeElement` as a parent for the frame for DOM search results)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +473,55 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         }
     }
 
+    _renderResultsForResourceTreeElement(resourceTreeElement, limit) {
```
Style: `{` should be on its own line

ditto below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +473,55 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        while (limit--)
+            resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));
+
+        this._didRenderResultsForResourceTreeElement(resourceTreeElement, matchObjects.length);
```
this is only ever called here, so i dont think we need a separate method for it

rcaliman-apple replied:
I prefer it as a separate method to communicate that it's a series of actions to be done _after_ rendering results:
- selecting the first result, if necessary
- managing the placeholder element

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +473,55 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        let incrementSize = Math.min(this._resultsIncrementSize, remainingSearchResults);
+        let remainingSize = remainingSearchResults > this._resultsIncrementSize ? remainingSearchResults : 0;
+
+        if (!placeholderElement) {
```
NIT: id actually flip this and put the stuff after inside an `if (placeholderElement) {` as omitting the `!` is one less operation (and also i kinda feel like we normally only have the negative pattern like this if we're about to `return` or set the variable to some know value for use after the `if`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeElement.js:
```diff
@@ -108,18 +109,6 @@ WI.TreeElement = class TreeElement extends WI.Object
-        return this._titleHTML;
-    }
-
-    set titleHTML(x)
```
huh i coulda sworn this was used at some point but i guess not anymore!  nice cleanup :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeElement.js:
```diff
@@ -241,11 +230,7 @@ WI.TreeElement = class TreeElement extends WI.Object
         if (!this._listItemNode)
             return;
 
-        if (!this._titleHTML && !this._title)
-            this._listItemNode.removeChildren();
```
i think this might still be useful if `.title = ""`

rcaliman-apple replied:
That case is covered by the existing logic which replaces all children with the empty text.

```js
if (typeof this._title === "string")
    this._listItemNode.textContent = this._title;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +473,55 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        placeholderElement.incrementSize = incrementSize;
+        placeholderElement.remainingSize = remainingSize;
+
+        resourceTreeElement.moveChild(placeholderElement, resourceTreeElement.children.length - 1);
```
is there a reason to do this instead of just `removeChild` and then `appendChild`?  is it just to avoid an additional call to `onattach`?

even simpler, why not just create an entirely new `WI.SearchResultsPlaceholderTreeElement` instead of saving it to `_searchResultsPlaceholderTreeElementForResourceTreeElement` since it doesn't really have any state (i.e. it just calls out to the delegate when clicked), which would also let us drop `set incrementSize` and `set remainingSize` and make them required arguments in the `constructor`

rcaliman-apple replied:
> even simpler, why not just create an entirely new WI.SearchResultsPlaceholderTreeElement

🤔 That's a good idea. I don't know why I didn't opt for that in the first place.

rcaliman-apple replied:
I remember! 

I opted for `this._searchResultsPlaceholderTreeElementForResourceTreeElement` as a `WeakMap` because I didn't want to store the placeholder tree element as a member of `WI.ResourceTreeElement` and `WI.ScriptTreeElement` or even lower in `WI.TreeElement`. It felt like polluting the scope with a specialization used now just for search sidebar panel. 

(Arguably, we could make this "load more" behavior a general trait of tree outline / tree element since it might apply beyond search results, but I didn't have a justification make a change that widespread just yet).

`WI.ScriptTreeElement` and `WI.ScriptTreeElement` instances are already keys for the lists of search match objects. I wanted to reuse that for consistency.

So the `new WI.SearchResultsPlaceholderTreeElement` has to belong somewhere to be retrieved.


~Removing and creating it anew in seems equivalent, but it causes the tree outline to fire `WI.TreeOutline.Event.ElementRemoved` and `WI.TreeOutline.Event.ElementAdded` and I wanted to avoid potential side-effects of that in the future (the events don't currently have handlers in the search sidebar panel).~

~If you don't have a strong preference, I'd keep this as-is.~

Update: Nevermind, I like this idea of removing and inserting a new tree element. It simplifies the logic. The `WeakMap` has to stay, though.

dcrousso replied:
since this entire PR is all about performance, id definitely lean towards avoiding having a `WeakMap` for this since that will consume arguably unnecessary memory (and in the case of a short search string that could be a _lot_)

the fact that there are no handlers for those events right now also makes me more in favor of just `new WI.SearchResultsPlaceholderTreeElement` each time we need to render one (and as a counterargument i feel like `moveChild` could make the logic of `WI.TreeOutline` more complicated in the future as until that there's been no need for a "insert without removing")

overall i think it would just make a lot of this logic much simpler

rcaliman-apple replied:
Agreed, I went with the approach to destroy and re-create a `WI.SearchResultsPlaceholderTreeElement`.

The `WeakMap` here holds a weak reference only between the top-level script / resource tree element and the _eventual_ placeholder element. It's not between individual search result items, of which there can be a lot. So any overhead is minimal. 

Either way a reference to the placeholder has to live somewhere, in a `WeakMap` or on the script / resource tree element itself, so we can retrieve it and remove it after rendering more/all results.

dcrousso reviewed (commented):
the functionality of this looks awesome

just a few comments that should hopefully make things simpler :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,81 @@
+    ondetach()
+    {
+        if (this._showMoreButtonElement)
+            this._showMoreButtonElement.removeEventListener("click", this._boundHandleShowMoreButtonClicked);
```
we dont normally call `removeEventListener` on DOM nodes since once their removed from the DOM they should never be invoked again

in this case, once `ondetach` is called we know that it doesn't exist in the DOM anymore, so the `"click"` event listener is already effectively removed (at least as far as we care)

ditto for the other one below (meaning we can entirely drop `ondetach`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js:
```diff
@@ -0,0 +1,81 @@
+        this._remainingSize = options.remainingSize ?? 0;
+
+        this._boundHandleShowMoreButtonClicked = this._handleShowMoreButtonClicked.bind(this);
+        this._boundHandleShowAllButtonClicked = this._handleShowAllButtonClicked.bind(this);
```
given below we shouldn't need to even save these to properties anymore

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -349,6 +341,33 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        console.assert(resourceTreeElement instanceof WI.ResourceTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);
+
+        this._renderResultsForResourceTreeElement(resourceTreeElement, limit);
+        this._renderResultsForResourceTreeElement(resourceTreeElement, limit);
```
oops?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +476,49 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+            }
+        }
+
+        let placeholderElement = this._searchResultsPlaceholderTreeElementForResourceTreeElement.get(resourceTreeElement);
```
we can completely remove `_searchResultsPlaceholderTreeElementForResourceTreeElement` if we pass the `searchResultsPlaceholderTreeElement` to this method instead of the `resourceTreeElement`

as such, id also probably rename this to be `_renderResultsForSearchResultsPlaceholderTreeElement`

rcaliman-apple replied:
Are you suggesting to use `WI.SearchResultsPlaceholderTreeElement` as the key for `_pendingSearchMatchObjectsForResourceTreeElement` WeakMap? (And rename it, of course).

That is possible, with a bit of juggling 🤔:

To be able to pass `searchResultsPlaceholderTreeElement` here, we'd have to create it preemptively in the callbacks from `performSearch()` to be able to render the first batch of results.

We don't know how many results there are at creation time, so we'd initialize the placeholder with `incrementSize = 0` and `remainingSize = 0`. We'd also want to avoid attaching it to the tree outline. This means checking later for `_didRenderResultsForResourceTreeElement()` before attempting to remove it.

We'd also have to take care and replace the `WeakMap` key with the new placeholder element after removing / re-creating it.

rcaliman-apple replied:
> Are you suggesting to use `WI.SearchResultsPlaceholderTreeElement` as the key for `_pendingSearchMatchObjectsForResourceTreeElement` WeakMap? (And rename it, of course).

Actually, that's not necessary. The key can remain the instance of resource / script tree element.

I tried a patch with the suggestion to use the placeholder element as an argument for the method to render results. I'm not convinced this is cleaner than the current proposal with an extra lightweight `WeakMap`:
- it's additional boilerplate in callbacks in `performSearch()` which I tried to reduce
- it requires attaching the placeholder tree element to the resource tree element to be able to reference the other later in `_renderResultsForPlaceholderTreeElement()` and `_didRenderResultsForPlaceholderTreeElement()` 
- it requires **always** removing the initial placeholder, with all incurred costs for doing that in `WI.TreeOutline`, after rendering the first results, regardless how many there are.

```diff
diff --git a/Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js b/Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js
index d0700ecf6e5e..daa0f709c777 100644
--- a/Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js
+++ b/Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js
@@ -61,7 +61,6 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
 
         this._omitMovingFocusToFirstResult = false;
         this._pendingSearchMatchObjectsForResourceTreeElement = new WeakMap;
-        this._searchResultsPlaceholderTreeElementForResourceTreeElement = new WeakMap;
     }
 
     // Public
@@ -136,7 +135,6 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         this.contentTreeOutline.removeChildren();
         this.contentBrowser.contentViewContainer.closeAllContentViews();
         this._pendingSearchMatchObjectsForResourceTreeElement = new WeakMap;
-        this._searchResultsPlaceholderTreeElementForResourceTreeElement = new WeakMap;
 
         let createSearchingPlaceholder = () => {
             let searchingPlaceholder = WI.createMessageTextView("");
@@ -198,7 +196,13 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
             if (!resource)
                 return;
 
+            const incrementSize = 0;
+            const remainingSize = 0;
+            let placeholderTreeElement = new WI.SearchResultsPlaceholderTreeElement(this, incrementSize, remainingSize);
+            placeholderTreeElement.selectable = false;
+
             let resourceTreeElement = this._searchTreeElementForResource(resource);
+            resourceTreeElement.appendChild(placeholderTreeElement);
 
             let searchMatchObjects = [];
             this._pendingSearchMatchObjectsForResourceTreeElement.set(resourceTreeElement, searchMatchObjects);
@@ -209,7 +213,7 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
                 });
             }
 
-            this._renderResultsForResourceTreeElement(resourceTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
+            this._renderResultsForPlaceholderTreeElement(placeholderTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
         };
 
         let resourcesCallback = ({result}) => {
@@ -241,7 +245,13 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
             if (!result || !result.length)
                 return;
 
-            var scriptTreeElement = this._searchTreeElementForScript(script);
+            const incrementSize = 0;
+            const remainingSize = 0;
+            let placeholderTreeElement = new WI.SearchResultsPlaceholderTreeElement(this, incrementSize, remainingSize);
+            placeholderTreeElement.selectable = false;
+
+            let scriptTreeElement = this._searchTreeElementForScript(script);
+            scriptTreeElement.appendChild(placeholderTreeElement);
 
             let searchMatchObjects = [];
             this._pendingSearchMatchObjectsForResourceTreeElement.set(scriptTreeElement, searchMatchObjects);
@@ -252,7 +262,7 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
                 });
             }
 
-            this._renderResultsForResourceTreeElement(scriptTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
+            this._renderResultsForPlaceholderTreeElement(placeholderTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
         };
 
         let domCallback = ({searchId, resultCount}) => {
@@ -268,7 +278,7 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
                 if (this._domSearchIdentifier !== searchId)
                     return;
 
-                let resourceTreeElementsWithSearchMatches = new Set;
+                let placeholderTreeElements = new Set;
 
                 for (let nodeId of nodeIds) {
                     let domNode = WI.domManager.nodeForId(nodeId);
@@ -301,12 +311,19 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
 
                     if (searchMatchObjects.length) {
                         this._pendingSearchMatchObjectsForResourceTreeElement.set(resourceTreeElement, searchMatchObjects);
-                        resourceTreeElementsWithSearchMatches.add(resourceTreeElement);
+
+                        const incrementSize = 0;
+                        const remainingSize = 0;
+                        let placeholderTreeElement = new WI.SearchResultsPlaceholderTreeElement(this, incrementSize, remainingSize);
+                        placeholderTreeElement.selectable = false;
+
+                        resourceTreeElement.appendChild(placeholderTreeElement);
+                        placeholderTreeElements.add(placeholderTreeElement);
                     }
                 }
 
-                for (let resourceTreeElement of resourceTreeElementsWithSearchMatches)
-                    this._renderResultsForResourceTreeElement(resourceTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
+                for (let placeholderTreeElement of placeholderTreeElements)
+                    this._renderResultsForPlaceholderTreeElement(placeholderTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
             };
 
             countPromise(target.DOMAgent.getSearchResults(searchId, 0, resultCount), domSearchResultsCallback);
@@ -361,10 +378,7 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
 
     searchResultsPlaceholderLoadMoreResults(searchResultsPlaceholderTreeElement, limit)
     {
-        let resourceTreeElement = searchResultsPlaceholderTreeElement.parent;
-        console.assert(resourceTreeElement instanceof WI.ResourceTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);
-
-        this._renderResultsForResourceTreeElement(resourceTreeElement, limit);
+        this._renderResultsForPlaceholderTreeElement(searchResultsPlaceholderTreeElement, limit);
     }
 
     // Private
@@ -475,8 +489,11 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         }
     }
 
-    _renderResultsForResourceTreeElement(resourceTreeElement, limit)
+    _renderResultsForPlaceholderTreeElement(placeholderTreeElement, limit)
     {
+        let resourceTreeElement = placeholderTreeElement.parent;
+        console.assert(resourceTreeElement instanceof WI.resourceTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);
+
         let matchObjects = this._pendingSearchMatchObjectsForResourceTreeElement.get(resourceTreeElement);
 
         limit = Math.min(matchObjects.length, limit);
@@ -484,25 +501,15 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         while (limit--)
             resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));
 
-        this._didRenderResultsForResourceTreeElement(resourceTreeElement, matchObjects.length);
+        this._didRenderResultsForPlaceholderTreeElement(placeholderTreeElement, matchObjects.length);
     }
 
-    _didRenderResultsForResourceTreeElement(resourceTreeElement, remainingSearchResults)
+    _didRenderResultsForPlaceholderTreeElement(placeholderTreeElement, remainingSearchResults)
     {
-        if (!this.contentTreeOutline.selectedTreeElement) {
-            let treeElement = resourceTreeElement.children.firstValue;
-            console.assert(treeElement instanceof WI.SearchResultTreeElement, treeElement);
-            if (treeElement) {
-                const selectedByUser = true;
-                treeElement.revealAndSelect(this._omitMovingFocusToFirstResult, selectedByUser);
-            }
-        }
+        let resourceTreeElement = placeholderTreeElement.parent;
+        console.assert(resourceTreeElement instanceof WI.SearchResultsPlaceholderTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);
 
-        let placeholderElement = this._searchResultsPlaceholderTreeElementForResourceTreeElement.get(resourceTreeElement);
-        if (placeholderElement) {
-            resourceTreeElement.removeChild(placeholderElement);
-            this._searchResultsPlaceholderTreeElementForResourceTreeElement.delete(resourceTreeElement);
-        }
+        resourceTreeElement.removeChild(placeholderTreeElement);
 
         if (!remainingSearchResults)
             return;
@@ -510,12 +517,19 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         let incrementSize = Math.min(WI.SearchSidebarPanel._resultsIncrementSize, remainingSearchResults);
         let remainingSize = remainingSearchResults > WI.SearchSidebarPanel._resultsIncrementSize ? remainingSearchResults : 0;
 
-        placeholderElement = new WI.SearchResultsPlaceholderTreeElement(this, incrementSize, remainingSize);
-        placeholderElement.selectable = false;
+        placeholderTreeElement = new WI.SearchResultsPlaceholderTreeElement(this, incrementSize, remainingSize);
+        placeholderTreeElement.selectable = false;
 
-        resourceTreeElement.appendChild(placeholderElement);
+        resourceTreeElement.appendChild(placeholderTreeElement);
 
-        this._searchResultsPlaceholderTreeElementForResourceTreeElement.set(resourceTreeElement, placeholderElement);
+        if (!this.contentTreeOutline.selectedTreeElement) {
+            let treeElement = resourceTreeElement.children.firstValue;
+            console.assert(treeElement instanceof WI.SearchResultTreeElement, treeElement);
+            if (treeElement) {
+                const selectedByUser = true;
+                treeElement.revealAndSelect(this._omitMovingFocusToFirstResult, selectedByUser);
+            }
+        }
     }
 
     _contentChanged(event)
```

I mean, it works. But is it really that much better?

rcaliman-apple replied:
@dcrousso I lived with this for a while and it has grown on me. 

I updated the patch to drop the `WeakMap` and use the placeholder element as a argument for `_renderResultsForPlaceholderTreeElement()` and `_didRenderResultsForPlaceholderTreeElement()`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +476,49 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        limit = Math.min(matchObjects.length, limit);
+
+        while (limit--)
+            resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));
```
i just realized that we're potentially calling `shift()` a whole bunch of times

do we know what the performance of that is vs a single `splice()` after the `while`?

even better, is there a way to do this without modifying the array at all (i.e. keeping track of an index)?

rcaliman-apple replied:
Intriguing! I did some performance measurements and found no meaningful difference between this approach with `shift()` compared to iterating through the array via index and `splice()` one-time:

This is how I instrumented:

```diff
+        let DEBUGperfStart = performance.now();
+        let DEBUGOriginalLimit = limit;
         while (limit--)
             resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));
 
+        console.log(`Array.prototype.shift(): Took ${performance.now() - DEBUGperfStart}ms to iterate ${DEBUGOriginalLimit} match objects and create search results.`)
```

and 
```diff
-        while (limit--)
-            resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));
+        let DEBUGperfStart = performance.now();
+        let DEBUGOriginalLimit = limit;
+        for (let i = 0; i < limit; i++)
+            resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects[i]));
+
+        matchObjects = matchObjects.splice(0, limit);
+
+        console.log(`For loop + one-time Array.prototype.splice(): Took ${performance.now() - DEBUGperfStart}ms to iterate ${DEBUGOriginalLimit} match objects and create search results.`)
```


With `limit = 1000` (default is `100`)
<img width="1276" height="197" alt="Screenshot 2026-01-21 at 13 21 48" src="https://github.com/user-attachments/assets/1bc489c1-7bff-41dc-b1da-aaaf8c570ecf" />

<img width="1278" height="197" alt="Screenshot 2026-01-21 at 13 31 22" src="https://github.com/user-attachments/assets/a191c197-bc8b-495b-9eec-df3696389303" />

With `limit = 5000`
<img width="1279" height="197" alt="Screenshot 2026-01-21 at 13 24 06" src="https://github.com/user-attachments/assets/ff202a84-417c-40c8-8799-d261900330df" />
<img width="1279" height="197" alt="Screenshot 2026-01-21 at 13 32 43" src="https://github.com/user-attachments/assets/00ca31cb-9bf5-4fa4-891a-3c236a43ba6a" />


I was puzzled by the huge discrepancy between ~63ms and ~950ms for 5000 iterations the same code, but on different files. The was without the patch to optimize string operations in https://github.com/WebKit/WebKit/pull/56723. 

The very slow performance turned out to be isolated to the test CSS file which happened to have all 5000 results on a single line. Rebasing on top of that patch, the perf improved dramatically.

Before & After https://github.com/WebKit/WebKit/pull/56723 applied:
<img width="1279" height="201" alt="Screenshot 2026-01-21 at 13 39 33" src="https://github.com/user-attachments/assets/a8af8316-9d9d-4fcc-822a-209ce7d75187" />

<img width="1280" height="196" alt="Screenshot 2026-01-21 at 13 51 39" src="https://github.com/user-attachments/assets/9b9f48d6-5edc-4b10-94bd-7f6d2867a932" />

For kicks, I also tried `limit = 100_000`, an extreme case:
 
<img width="1012" height="46" alt="Screenshot 2026-01-21 at 14 07 40" src="https://github.com/user-attachments/assets/4c952beb-a527-43b7-a759-1e78cc1c805b" />
<img width="1014" height="46" alt="Screenshot 2026-01-21 at 14 00 53" src="https://github.com/user-attachments/assets/b3d9a612-3ec3-4d40-9e47-ea2354bef6e9" />

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +476,49 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        let incrementSize = Math.min(WI.SearchSidebarPanel._resultsIncrementSize, remainingSearchResults);
+        let remainingSize = remainingSearchResults > WI.SearchSidebarPanel._resultsIncrementSize ? remainingSearchResults : 0;
+
+        placeholderElement = new WI.SearchResultsPlaceholderTreeElement(this, {incrementSize, remainingSize});
```
now that we create a new one of these each time, should we require `incrementSize` and `remainingSize` (i.e. it seems kinda pointless to create one with either value as `0`)?

that would also let us drop the `if` inside `WI.SearchResultsPlaceholderTreeElement.prototype.onattach`

rcaliman-apple replied:
That's a good point. But I'll keep the `if` statements in `WI.SearchResultsPlaceholderTreeElement.prototype.onattach`:

- it's possible to have `incrementSize` and no `remainingSize` when the remaining number of results is lower than the increment. We don't want to show the "Show All" button.

- I'd like to follow-up with a little _delight_ and pick a threshold where "Show All" makes more sense than "Show X More". 
For example, when the increment in 100 but there are 103 results left to show. It's kind of clumsy to say "Show 3 More". Might be nicer to just say "Show All (103 More)" instead. In that case `incrementSize` would `0`, while `remainingSize` would be `103`.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TreeOutline.css:
```diff
@@ -91,6 +91,21 @@ body[dir=rtl] .tree-outline .item :is(.disclosure-button, .icon) {
     margin-inline-start: var(--tree-outline-icon-margin-start);
 }
 
+.tree-outline .item.placeholder-item > button {
```
this is no longer needed now, right?

rcaliman-apple replied:
Good catch. It's a leftover.

dcrousso reviewed (commented):
a few more simplifications and one performance concern about how we handle the not-yet-shown results

burg reviewed (commented):
I haven't dug into the code changes yet, but I'm confident that Devin has taken a good look.

Both of my questions are about dynamic updates to the search sidebar.

What happens if the "More" placeholder is showing and the user enters text into the filter sidebar at the bottom? I'd expect that the placeholder result count tally get updated. This is how Xcode behaves.

I have a second functionality question: what happens if a selected tree element gets pushed down the results list when the query changes to get more results (i.e., deleting characters)? Do we auto-remove the placeholder if the query results no longer are overflowing?

If you encountered any bugs along the way.. if you have advice for quickly testing that this change actually works (I.e., so an integration engineer can quickly verify), I'd advise that you include that info in the commit message.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -196,26 +188,28 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
                 return;
 
-            var resourceTreeElement = this._searchTreeElementForResource(resource);
+            let placeholderTreeElement = this._createSearchResultsPlaceholderTreeElement();
```
why are we creating this element only to likely immediately remove it in `_renderResultsForPlaceholderTreeElement` (since most resources dont have that many matches)?

can't we wait to create the `WI.SearchResultsPlaceholderTreeElement` until we know for sure that there's more than `WI.SearchSidebarPanel._resultsIncrementSize` results to show?

rcaliman-apple replied:
>why are we creating this element only to likely immediately remove it in _renderResultsForPlaceholderTreeElement (since most resources dont have that many matches)?

Because we wanted a way to be able to call `_renderResultsForPlaceholderTreeElement` both from callbacks in `performSearch()` and from the delegate `searchResultsPlaceholderLoadMoreResults()` using the placeholder.

> can't we wait to create the WI.SearchResultsPlaceholderTreeElement until we know for sure that there's more than WI.SearchSidebarPanel._resultsIncrementSize results to show?

That's what the [previous version of the patch did](https://github.com/WebKit/WebKit/compare/f89b071e443bc5c3a95675d8ca6ec719c582ca42..020ef670ca19d1ca807c1573d05a3900b45de5bc). This change was in response to your ask to remove the additional `WeakMap` that mapped placeholders to resource tree elements. 

Placeholders used to just be created when needed in  `_didRenderResultsForResourceTreeElement` after rendering the first batch of results in `_renderResultsForResourceTreeElement`

rcaliman-apple replied:
No longer relevant. See comment https://github.com/WebKit/WebKit/pull/55424#discussion_r2733378560

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -247,25 +241,20 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
                 return;
 
-            var scriptTreeElement = this._searchTreeElementForScript(script);
+            let placeholderTreeElement = this._createSearchResultsPlaceholderTreeElement();
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -295,38 +286,44 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+                    if (searchMatchObjects.length) {
+                        this._pendingSearchMatchObjectsForResourceTreeElement.set(resourceTreeElement, searchMatchObjects);
 
+                        let placeholderTreeElement = this._createSearchResultsPlaceholderTreeElement();
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -295,38 +286,44 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
                 }
+
+                for (let placeholderTreeElement of placeholderTreeElements)
+                    this._renderResultsForPlaceholderTreeElement(placeholderTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
```
is there a reason we do this in a separate loop at the end instead of when the `WI.SearchResultsPlaceholderTreeElement` is created a few lines above?

rcaliman-apple replied:
```
for (let nodeId of nodeIds) {
...
```

There's no explicit guarantee that all nodeIds will be from the same frame. 
Rendering after that loop means all pending results have been collected into their appropriate bucket.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +477,57 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+    _didRenderResultsForPlaceholderTreeElement(placeholderTreeElement, remainingSearchResults)
+    {
+        let resourceTreeElement = placeholderTreeElement.parent;
+        console.assert(resourceTreeElement instanceof WI.ResourceTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);
```
having to do this same exact code again is why i recommend merging this method with `_renderResultsForPlaceholderTreeElement` as then we can use the other local variable already created for this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +477,57 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+        let resourceTreeElement = placeholderTreeElement.parent;
+        console.assert(resourceTreeElement instanceof WI.ResourceTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);
+
+        resourceTreeElement.removeChild(placeholderTreeElement);
```
as mentioned above, this only needs to happen for the codepath from `searchResultsPlaceholderLoadMoreResults`

id recommend moving all the below code into a new `_createSearchResultsPlaceholderTreeElementIfNeeded` that's called from the other callsites in `performSearch`
```js
searchResultsPlaceholderLoadMoreResults(searchResultsPlaceholderTreeElement, limit)
{
    let resourceTreeElement = searchResultsPlaceholderTreeElement.parent;
    console.assert(resourceTreeElement instanceof WI.ResourceTreeElement || resourceTreeElement instanceof WI.ScriptTreeElement, resourceTreeElement);

    resourceTreeElement.removeChild(searchResultsPlaceholderTreeElement);

    let matchObjects = this._pendingSearchMatchObjectsForResourceTreeElement.get(resourceTreeElement);

    limit = Math.min(matchObjects.length, limit);
    while (limit--)
        resourceTreeElement.appendChild(new WI.SearchResultTreeElement(matchObjects.shift()));

    this._createSearchResultsPlaceholderTreeElementIfNeeded(resourceTreeElement, matchObjects);
}

/* ... */

_createSearchResultsPlaceholderTreeElementIfNeeded(resourceTreeElement, remainingSearchResults)
{
    if (!remainingSearchResults.length)
        return;

    let incrementSize = Math.min(WI.SearchSidebarPanel._resultsIncrementSize, remainingSearchResults.length);
    let remainingSize = remainingSearchResults.length > WI.SearchSidebarPanel._resultsIncrementSize ? remainingSearchResults.length : 0;

    let placeholderTreeElement = new WI.SearchResultsPlaceholderTreeElement(this, incrementSize, remainingSize);
    placeholderTreeElement.selectable = false;
    resourceTreeElement.appendChild(placeholderTreeElement);

    if (!this.contentTreeOutline.selectedTreeElement) {
        let treeElement = resourceTreeElement.children.firstValue;
        console.assert(treeElement instanceof WI.SearchResultTreeElement, treeElement);
        if (treeElement) {
            const selectedByUser = true;
            treeElement.revealAndSelect(this._omitMovingFocusToFirstResult, selectedByUser);
        }
    }
}
```

rcaliman-apple replied:
I'll hold off on this until you and @burg get a chance to look at the filtering proposal.

I believe `_renderResultsForPlaceholderTreeElement()` becomes an important requirement there to satisfy the cases for:
- filtering after a search 
- searching with a filter applied
- changing the filter multiple times on the same results

The placeholder element will potentially need to be added / removed independent of the delegate `searchResultsPlaceholderLoadMoreResults()`

rcaliman-apple replied:
@dcrousso I took your suggestion and removed `_didRenderResultsForPlaceholderTreeElement` and merged the code into the delegate `searchResultsPlaceholderLoadMoreResults`.  

As of the [latest changes](https://github.com/WebKit/WebKit/compare/66e194b1803718560678e93404ae28c32b020681..389e87686dfe216a969414065bc931b422068023), we're no longer creating placeholder elements unnecessarily.

 I think this addresses all of your review comments.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -196,26 +188,26 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
-            if (!resourceTreeElement.children.length)
-                this.contentTreeOutline.removeChild(resourceTreeElement);
+            this._renderResultsForResourceTreeElement(resourceTreeElement, WI.SearchSidebarPanel._resultsIncrementSize);
+            this._createSearchResultsPlaceholderTreeElementIfNeeded(resourceTreeElement, searchMatchObjects);
```
why are we not saving the result of `_renderResultsForResourceTreeElement` and passing it to `_createSearchResultsPlaceholderTreeElementIfNeeded`?

is it just because `searchMatchObjects` is the same value that's put into the `_pendingSearchMatchObjectsForResourceTreeElement`?

if so, that's a potential bug in the future if the implementation of `_renderResultsForResourceTreeElement` changes, so id recommend not doing that

dcrousso replied:
alternatively, we could change `_renderResultsForResourceTreeElement` to require the `searchMatchObjects` as an argument so that it doesn't have to re-fetch the value we already have locally from `_pendingSearchMatchObjectsForResourceTreeElement` (i.e. avoid an extra map lookup)

rcaliman-apple replied:
>why are we not saving the result of _renderResultsForResourceTreeElement and passing it to _createSearchResultsPlaceholderTreeElementIfNeeded?

It's amusing how I considered what you'd say and chose wrong 😅. I thought you'd discourage the creation of another variable when `searchMatchObjects` is right there.

But now, reading your feedback, I agree that it's error-prone to assume that the `searchMatchObjects` array which `_renderResultsForResourceTreeElement` mutates will be the same over time. I updated the patch to use the same pattern of using the output of `_renderResultsForResourceTreeElement` everywhere.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -349,6 +341,40 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+
+    matchTreeElementAgainstCustomFilters(treeElement)
+    {
+        return treeElement instanceof WI.SearchResultsPlaceholderTreeElement ? false : true;
```
```js
return !(treeElement instanceof WI.SearchResultsPlaceholderTreeElement);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +483,42 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
         }
     }
 
+    _renderResultsForResourceTreeElement(resourceTreeElement, limit)
```
NIT: we really should call something more generic like `sourceCodeTreeElement` since it can also be a `WI.ScriptTreeElement` (i.e. using "resource" is technically not entirely accurate)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
```diff
@@ -457,6 +483,42 @@ WI.SearchSidebarPanel = class SearchSidebarPanel extends WI.NavigationSidebarPan
+            return;
+
+        let incrementSize = Math.min(WI.SearchSidebarPanel._resultsIncrementSize, remainingSearchResults.length);
+        let remainingSize = remainingSearchResults.length > WI.SearchSidebarPanel._resultsIncrementSize ? remainingSearchResults.length : 0;
```
NIT: i kinda feel like it's odd for `remainingSize` to be set to `0` if `incrementSize` is greater than `0`.  what if instead we moved this logic inside `WI.SearchResultsPlaceholderTreeElement` since ultimately that's what should decide what buttons to show (i.e. the caller here should really only provide the total number of remaining things and how many to show incrementally with the decision about what buttons to show happening in `WI.SearchResultsPlaceholderTreeElement`)
```js
let placeholderTreeElement = new WI.SearchResultsPlaceholderTreeElement(this, WI.SearchSidebarPanel._resultsIncrementSize, remainingSearchResults.length);
```
then in `WI.SearchResultsPlaceholderTreeElement` we could have
```js
constructor(delegate, incrementCount, remainingCount)
{
    console.assert(this._delegate?.searchResultsPlaceholderLoadMoreResults, this._delegate);
    console.assert(incrementCount > 0, incrementCount);
    console.assert(remainingCount > 0, remainingCount);

    const title = "";
    super(title);

    this._delegate = delegate;

    this._incrementCount = Math.min(incrementCount, remainingCount);
    this._remainingCount = remainingCount > incrementCount ? remainingCount : 0;
}
```

rcaliman-apple replied:
I agree 👍

dcrousso reviewed (approved):
nice work!  just a few minor NITs and design suggestions.  thanks for iterating with me :)

rcaliman-apple committed:
Web Inspector: Performance issue with search tab and single character
rdar://49234522
https://bugs.webkit.org/show_bug.cgi?id=262158

Reviewed by Devin Rousso.

Web Inspector search works by first requesting a list of all resources
that have matches for a query. Then, iterating through each resource,
it gets lines of text where matches are found. Finally the same search
query regex is applied to each line to get individual search result matches.

Before this patch, a `WI.SearchResultTreeElement` was created and attached
for each match, regardless of how many. This was expensive for a few reasons:

- for each result, a few substring operations are run on the line of text
to provide context around the individual match. On minified files,
this means working on almost the entire contents of the file.

- attaching the tree element to the tree outline creates a DOM structure for each element,
without necessarily attaching it to the DOM (tree outline virtualization handles that),
adds event handlers, creates relationships with sibling tree elements, etc.
All this work is done regardless if the tree element will ever be seen.

For queries with a very large number of results (tens or hundreds of thousands),
there's a lot unnecessary work done before even showing the first few results.
This work occurs on the UI thread and blocks interaction within Web Inspector.

From a UX perspective, a user overwhelmed with a large volume of search results
is more likely to want to refine their query than see all of the results.

This patch reduces the amount of work done:
- for each resource, show only the first 100 results by default.

- provide UI controls per resource to load more results or all results.
This allows a user to intentionally request to show a known large number of results
with the implied acknowledgement that the action might cause a slowdown.
This follows a pattern used elsewhere in Web Inspector: showing contents of Arrays,
Sets, Maps, or DataGrids with many entries.

- other piecemeal optimizations:
-- event delegation for handling double-click on individual search result tree elements

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.css: Added.
(.tree-outline .item.search-results-placeholder > button):
(&:first-of-type):
* Source/WebInspectorUI/UserInterface/Views/SearchResultsPlaceholderTreeElement.js: Added.
(WI.SearchResultsPlaceholderTreeElement):
(WI.SearchResultsPlaceholderTreeElement.prototype.onattach):
(WI.SearchResultsPlaceholderTreeElement.prototype._handleShowMoreButtonClicked):
(WI.SearchResultsPlaceholderTreeElement.prototype._handleShowAllButtonClicked):
* Source/WebInspectorUI/UserInterface/Views/SearchSidebarPanel.js:
(WI.SearchSidebarPanel):
(WI.SearchSidebarPanel.prototype.performSearch):
(WI.SearchSidebarPanel.prototype.attached):
(WI.SearchSidebarPanel.prototype.detached):
(WI.SearchSidebarPanel.prototype.matchTreeElementAgainstCustomFilters):
(WI.SearchSidebarPanel.prototype.searchResultsPlaceholderLoadMoreResults):
(WI.SearchSidebarPanel.prototype._renderResultsForSourceCodeTreeElement):
(WI.SearchSidebarPanel.prototype._createSearchResultsPlaceholderTreeElementIfNeeded):
(WI.SearchSidebarPanel.prototype.performSearch.createTreeElementForMatchObject): Deleted.
(WI.SearchSidebarPanel.prototype._treeElementDoubleClick): Deleted.
* Source/WebInspectorUI/UserInterface/Views/TreeElement.js:
(WI.TreeElement.prototype._setListItemNodeContent):
(WI.TreeElement.prototype.get titleHTML): Deleted.
(WI.TreeElement.prototype.set titleHTML): Deleted.
Drive-by clean-up of unused `titleHTML`

Canonical link: https://commits.webkit.org/306429@main
